### DISCUSSION TITLE
What function or method takes searchState and converts it to requests for api call?

### DISCUSSION THREAD

**Discussion by hadminh** [algolia/instantsearch#5627] [General] [answered]
What function converts searchState to http_requests?

```js
const searchState= {
    "sortBy": "hotels_index",
    "page": "1",
    "refinementList": {
        "Municipality": [
            "London"
        ]
    },
    "query": ""
}
```

```js
const http_requests = [
    {
        "indexName": "hotels_index",
        "params": {
            "facetFilters": [
                [
                    "Municipality:London"
                ]
            ],
            "facets": [
                "Municipality"
            ],
            "highlightPostTag": "</ais-highlight-0000000000>",
            "highlightPreTag": "<ais-highlight-0000000000>",
            "hitsPerPage": 10,
            "maxValuesPerFacet": 1000,
            "page": 0,
            "query": "",
        }
    },
]
```

**Comment by hadminh** [comment]
I think the answer is [here](https://github.com/algolia/algoliasearch-helper-js/blob/6a36194d425b2446c6406e74df75c87bf5028454/src/requestBuilder.js#L16)

**Comment by hadminh** [comment, marked as answer]
```js
const algoliasearch = require("algoliasearch");
const algoliasearchHelper = require("algoliasearch-helper");
const requestBuilder = require("algoliasearch-helper/src/requestBuilder");
const client = algoliasearch("YourApplicationID", "YourSearchOnlyAPIKey");
const index = client.initIndex("YourIndexName");

const searchState = {
  range: {
    price: {
      min: 20,
      max: 3000,
    },
  },
  refinementList: {
    fruits: ["lemon", "orange"],
  },
  menu: {
    brands: "Sony",
  },
  query: "ora",
  page: 2,
};

const helper = algoliasearchHelper(client, index.indexName, {
  disjunctiveFacets: ["fruits"],
  facets: ["brands"],
  hitsPerPage: 10,
  page: searchState.page,
  numericFilters: [`price >= ${searchState.range.price.min}`, `price <= ${searchState.range.price.max}`],
  query: searchState.query,
});

if (searchState.refinementList && searchState.refinementList.fruits) {
  helper.addDisjunctiveFacetRefinement("fruits", searchState.refinementList.fruits);
}

if (searchState.menu && searchState.menu.brands) {
  helper.addFacetRefinement("brands", searchState.menu.brands);
}

const request = helper.state.getQueryParams({});

console.log(request);
const queries = requestBuilder._getQueries(index, helper.state);

console.log(queries);

```

**Reply by Haroenv** [reply]
That's correct, but the translation from searchState to SearchParameters actually happens here: https://github.com/algolia/instantsearch/blob/162341a75a81b97fbcc14107114437d8d92a9d18/packages/react-instantsearch-core/src/core/createInstantSearchManager.js#L137-L211

**Reply by hadminh** [reply]
Thank you for offering a better solution than my original one that relied on algoliasearch-helper.
Have you thought of providing a lower level api using a general standard state management tool like zustand/redux etc, instead of using algoliasearch-helper?

**Reply by Haroenv** [reply]
The searchState (UiState in modern React InstantSearch Hooks versions) should really be controlled by InstantSearch, making a change to it without notifying the widgets that actually own the parameters would not work efficiently.

However, useInstantSearch has a `setUiState` function that can be called freely, as well as the InstantSearch component having a onStateChange option (can also be done with a middleware). Both of those solutions are assuming React InstantSearch Hooks, as that's the modern version of React InstantSearch :)

What are you trying to achieve exactly by controlling the state?

**Reply by hadminh** [reply]
Almost every year, a new framework emerges, such as Remix, Svelt, SolidJS, React, Angular, or Vue. These frameworks continuously evolve, with Next.js being a prime example of significant improvement between versions 12 and 13. Considering this, I believe it's a wise choice to build the entire UI state within Zustand. What sets Zustand apart is its independence from the "DOM" and its framework-friendly nature. This means that even as the frameworks receive updates, I can still utilize Zustand independently, ensuring I stay up-to-date with the latest framework versions. This becomes especially important as these frameworks eventually phase out older versions and drop support.
